test/Microsoft.ML.Tests/TrainerEstimators/MatrixFactorizationTests.cs

```diff
@@ -0,0 +1,115 @@
+
+                // Create reader for both of training and test data sets
+                var reader = new TextLoader(env, GetMfTrivialLoaderArgs(labelColumnName, userColumnName, itemColumnName));
+
+                // Read training data as an IDataView object
+                var data = reader.Read(new MultiFileSource(GetDataPath(TestDatasets.trivialMatrixFactorization.trainFilename)));
+
+                // Create a pipeline with a single operator.
+                var pipeline = new MatrixFactorizationTrainer(env, labelColumnName, userColumnName, itemColumnName, 
+                    advancedSettings:s=>
+                    {
+                        s.NumIterations = 3;
+                        s.NumThreads = 1; // To eliminate randomness, # of threads must be 1.
+                        s.K = 7;
+                    });
+
+                // Train a matrix factorization model.
+                var model = pipeline.Fit(data);
+
+                // Read the test data set as an IDataView
+                var testData = reader.Read(new MultiFileSource(GetDataPath(TestDatasets.trivialMatrixFactorization.testFilename)));
+
+                // Apply the trained model to the test set
+                var prediction = model.Transform(testData);
+
+                // Retrieve label column's index from the test IDataView
+                testData.Schema.TryGetColumnIndex(labelColumnName, out int labelColumnId);
+
+                // Retrieve score column's index from the IDataView produced by the trained model
+                prediction.Schema.TryGetColumnIndex(scoreColumnName, out int scoreColumnId);
+
+                // Go through the whole test set and calculate the expected statistics.
+                double error = 0;
+                long count = 0;
+                using (var cursor = testData.GetRowCursor(c => c == labelColumnId))
+                using (var scoreCursor = prediction.GetRowCursor(c => c == scoreColumnId))
+                {
+                    float label = 0;
+                    float score = 0;
+                    ValueGetter<float> labelGetter = RowCursorUtils.GetGetterAs<float>(NumberType.R4, cursor, labelColumnId);
+                    ValueGetter<float> scoreGetter = RowCursorUtils.GetGetterAs<float>(NumberType.R4, scoreCursor, scoreColumnId);
+                    while (cursor.MoveNext() && scoreCursor.MoveNext())
+                    {
+                        labelGetter(ref label);
+                        scoreGetter(ref score);
+                        error += (score - label) * (score - label);
+                        ++count;
+                    }
+                }
+                error /= count;
+                double tolerance = System.Math.Pow(10, -DigitsOfPrecision);
+                double expectedError = 0.61528733643754685;
+                double relativeError = System.Math.Abs(error - expectedError) / System.Math.Abs(expectedError);
+                Assert.True(relativeError < tolerance);
```

Comment:
Can we use `Assert.InRange` instead?

Reply:
I switched to Assert.InRange for a while and then finally use Assert.True because different platforms require different ranges.

---
In reply to: [226721494](https://github.com/dotnet/machinelearning/pull/1263#discussion_r226721494) [](ancestors = 226721494)